crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -312,69 +317,99 @@ impl RecursorDnsHandle {
-                self.resolve(a_query, request_time, false).boxed()
-            });
+            debug!("need glue for {zone}");
 
-            let aaaa_resolves = need_ips_for_names.iter().take(1).map(|name| {
-                let aaaa_query = Query::query(name.0.clone(), RecordType::AAAA);
-                self.resolve(aaaa_query, request_time, false).boxed()
-            });
+            let mut resolve_futures = vec![];
+            need_ips_for_names
+                .iter()
+                .filter(|name| !crate::is_subzone(&zone, &name.0))
+                .take(1)
+                .for_each(|name| {
+                    for rec_type in [RecordType::A, RecordType::AAAA] {
+                        resolve_futures.push(
+                            self.resolve(
+                                Query::query(name.0.clone(), rec_type),
+                                request_time,
+                                self.security_aware,
+                            )
+                            .boxed(),
+                        );
+                    }
+                });
 
-            let mut a_resolves: Vec<_> = a_resolves.chain(aaaa_resolves).collect();
-            while !a_resolves.is_empty() {
-                let (next, _, rest) = select_all(a_resolves).await;
-                a_resolves = rest;
+            while !resolve_futures.is_empty() {
+                let (next, _, rest) = select_all(resolve_futures).await;
+                resolve_futures = rest;
 
                 match next {
-                    Ok(response) => {
-                        debug!("A or AAAA response: {:?}", response);
-                        let ips = response.iter().filter_map(RData::ip_addr);
+                    Ok(res) => {
+                        let ips = res.iter().filter_map(RData::ip_addr).collect();
+                        super::add_config_group_recs(ips, &mut config_group);
+                    }
+                    Err(e) => warn!("unable to resolve NS: {e}"),
+                }
+            }
+        }
 
-                        for ip in ips {
-                            let udp =
-                                NameServerConfig::new(SocketAddr::from((ip, 53)), Protocol::Udp);
-                            let tcp =
-                                NameServerConfig::new(SocketAddr::from((ip, 53)), Protocol::Tcp);
+        // If we still have no NS records, try to query the parent zone for child NS servers
+        // Note that while this section looks very similar to the previous section, there is
+        // a very important difference: the use of lookup to resolve NS addresses, vs resolve
+        // in the previous section.  Using resolve here will cause an infinite loop for these
+        // nameservers. Using lookup with nameserver_pool in the previous section would almost
+        // always cause resolution failures.
+        if config_group.is_empty() && !need_ips_for_names.is_empty() {
```

Comment:
Style nit: invert this condition and return early?

Reply:
I looked at this - it would result in some duplication in constructing the GenericNameServerPool.